src/audio/copier/copier.c

```diff
@@ -52,11 +52,30 @@ static pcm_converter_func get_converter_func(const struct ipc4_audio_format *in_
 					     enum ipc4_gateway_type type,
 					     enum ipc4_direction_type);
 
+static uint32_t bitmask_to_nibble_channel_map(uint8_t bitmask)
+{
+	int i, channel_count = 0;
+	uint32_t nibble_map = 0;
+
+	for (i = 0; i < 8; i++)
+		if (bitmask & (1 << i)) {
+			nibble_map |= i << (channel_count * 4);
```

Comment:
how does this work for i == 0 ?

Reply:
It does work. nibble_map is a channel map in which each channel is represented by 4 bits (total 8 channels in uint32_t) as value from 0 to 7 and a special value 0xf means no channel (mute). For example, map for stereo would be 0x00000010.
Here for i == 0 (channel 0) value 0 will be ORed, i.e., OR will do nothing. However, since nibble_map is initialized with 0, the result will be 0 and this is what we want (channel 0 is represented as 0x0 nibble).

Reply:
so if bitmask is 0x1, nibble_map should be 0xFFFFFFF0?

Reply:
> so if bitmask is 0x1, nibble_map should be 0xFFFFFFF0?

Yes.

Reply:
Can this function be simplified then as 
```
static uint32_t bitmask_to_nibble_channel_map(uint8_t bitmask)
{
	uint32_t nibble_map;
	int channel_count = popcount(bitmask);
	int i;
	
	nibble_map = GEN_MASK(31, channel_count);
	for (i = 0; i < channel_count; i++)
		nibble_map |= i << (i * 4);

	return nibble_map;
}
````

Reply:
No. Bitmask could be like 00000101 binary, meaning two channels with number 0 and 2. This potentially could be later multiplexed with some other channels coming from other gateway to create, for example, a 4 channel stream with channels 0,1,2,3.

Reply:
So irrespective of the bitmask 0b11 or 0b101, the nibble mask remains the same 0xFFFFFF10?

Reply:
No, it are 0xFFFFFF10 and 0xFFFFFF20 respectively.